Sources/Plasma/CoreLib/hsStream.cpp

```diff
@@ -660,88 +624,96 @@ uint32_t  plReadOnlySubStream::GetEOF()
 
 //////////////////////////////////////////////////////////////////////////////////////
 
-#define kRAMStreamChunkSize     1024
-
-hsRAMStream::hsRAMStream() : fAppender(1, kRAMStreamChunkSize)
-{
-    fIter.ResetToHead(&fAppender);
-}
-
-hsRAMStream::hsRAMStream(uint32_t chunkSize) : fAppender(1, chunkSize)
-{
-    fIter.ResetToHead(&fAppender);
-}
-
-hsRAMStream::~hsRAMStream()
-{
-}
-
-void hsRAMStream::Reset()
-{
-    fBytesRead = 0;
-    fPosition = 0;
-
-    fAppender.Reset();
-    fIter.ResetToHead(&fAppender);
-}
-
 bool hsRAMStream::AtEnd()
 {
-    return (fBytesRead >= fAppender.Count() * fAppender.ElemSize());
+    return (fPosition >= fVector.size());
 }
 
 uint32_t hsRAMStream::Read(uint32_t byteCount, void * buffer)
 {
-    if (fBytesRead + byteCount > fAppender.Count() * fAppender.ElemSize())
-        byteCount = (fAppender.Count() * fAppender.ElemSize()) - fBytesRead;
+    if (fPosition + byteCount > fVector.size()) {
+//      hsStatusMessageF("Reading past end of hsRAMStream (read %u of %u requested bytes)", fVector.size() - fPosition, byteCount);
+        byteCount = fVector.size() - fPosition;
+    }
+    
+    memcpy(buffer, fVector.data() + fPosition, byteCount);
 
-    fBytesRead += byteCount;
     fPosition += byteCount;
 
-    fIter.Next(byteCount, buffer);
-
     return byteCount;
 }
 
 uint32_t hsRAMStream::Write(uint32_t byteCount, const void* buffer)
 {
-    fPosition += byteCount;
+    size_t spaceUntilEof = fVector.size() - fPosition;
+    if (byteCount <= spaceUntilEof) {
+        memcpy(fVector.data() + fPosition, buffer, byteCount);
+    } else {
+        memcpy(fVector.data() + fPosition, buffer, spaceUntilEof);
+        auto buf = static_cast<const uint8_t*>(buffer);
+        fVector.insert(fVector.end(), buf + spaceUntilEof, buf + byteCount);
+    }
 
-    fAppender.PushTail(byteCount, buffer);
+    fPosition += byteCount;
 
     return byteCount;
 }
 
 void hsRAMStream::Skip(uint32_t deltaByteCount)
 {
     fPosition += deltaByteCount;
-    fIter.Next(deltaByteCount, nullptr);
 }
 
 void hsRAMStream::Rewind()
 {
-    fBytesRead = 0;
     fPosition = 0;
-    fIter.ResetToHead(&fAppender);
+}
+
+void hsRAMStream::FastFwd()
+{
+    fPosition = fVector.size();
 }
 
 void hsRAMStream::Truncate()
 {
-    Reset();
+    fVector.resize(fPosition);
```

Comment:
These two things are not equal...  The old `hsRAMStream::Truncate()` would always reset the stream to 0 size.  Did you check everywhere Truncate was called to see which behavior was expected?

Reply:
Oh, good point, I didn't even notice that. Both `hsUNIXStream` and the former `hsVectorStream` implement `Truncate` as "delete all data after current seek position", so I would say the old `hsRAMStream::Truncate` was simply incorrect.

I'll have to check if anything relied on the old `hsRAMStream` behavior...

Reply:
Okay, thankfully `hsStream::Truncate` is basically never used. It's only called three times in all of Plasma, and two of those calls are inside other `Truncate` implementations. The one "real" call was on a `hsRAMStream`, *of course*, so I changed it to `Reset`.